File: app/src/main/java/uca/esi/manual/models/labs/TorsionLab.java
package uca.esi.manual.models.labs;


import android.os.Parcel;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * The type Torsion lab.
 */
public class TorsionLab extends BaseLab {

    /**
     * The constant CREATOR.
     */
    public static final Creator<BaseLab> CREATOR = new Creator<BaseLab>() {

        @Override
        public BaseLab createFromParcel(Parcel source) {
            return new TorsionLab(source);  //using parcelable constructor
        }

        @Override
        public BaseLab[] newArray(int size) {
            return new BaseLab[size];
        }
    };


    private static final int[] LOADS_TORSION = {5, 10, 15, 20};
    private static final float[] LOAD_MOMENT = {.5f, 1.0f, 1.5f, 2.0f};
    private static final float[] AMPLI_VALUES = {-.035f, -.073f, -.112f, -.15f};
    private int weights = 0;
    private int errWeights = 0;
    private float ampliTeo = 0;
    private float ampliExp = 0;

    /**
     * Instantiates a new Torsion lab.
     *
     * @param userId the user id
     */
    public TorsionLab(String userId) {
        super(userId, LabType.TORSION);
        setData();
    }

    /**
     * Instantiates a new Torsion lab.
     *
     * @param in the in
     */
    public TorsionLab(Parcel in) {
        super(in);
        String[] data = in.createStringArray();
        if (data != null) {
            this.weights = Integer.parseInt(data[0]);
            this.errWeights = Integer.parseInt(data[1]);
            this.ampliTeo = Float.parseFloat(data[2]);
            this.ampliExp = Float.parseFloat(data[3]);
        }
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeStringArray(torsionToStringList().toArray(new String[0]));
    }

    private List<String> torsionToStringList() {
        List<String> data = new ArrayList<>();
        data.add(String.valueOf(weights));
        data.add(String.valueOf(errWeights));
        data.add(String.valueOf(ampliTeo));
        data.add(String.valueOf(ampliExp));

        return data;
    }

    @Override
    protected void setData() {
        int index = (new Random().nextInt(LOADS_TORSION.length));
        this.weights = LOADS_TORSION[index];
        setValTeo(LOAD_MOMENT[index]);
        this.ampliTeo = AMPLI_VALUES[index];
    }

    @Override
    public String getData() {
        return String.valueOf(weights);
    }

    /**
     * Gets weights.
     *
     * @return the weights
     */
    public int getWeights() {
        return weights;
    }

    /**
     * Sets weights.
     *
     * @param weights the weights
     */
    public void setWeights(int weights) {
        this.weights = weights;
    }

    /**
     * Gets err weights.
     *
     * @return the err weights
     */
    public int getErrWeights() {
        return errWeights;
    }

    /**
     * Sets err weights.
     *
     * @param errWeights the err weights
     */
    public void setErrWeights(int errWeights) {
        this.errWeights = errWeights;
    }

    /**
     * Gets ampli teo.
     *
     * @return the ampli teo
     */
    public float getAmpliTeo() {
        return ampliTeo;
    }

    /**
     * Gets ampli exp.
     *
     * @return the ampli exp
     */
    public float getAmpliExp() {
        return ampliExp;
    }

    /**
     * Sets ampli exp.
     *
     * @param ampliExp the ampli exp
     */
    public void setAmpliExp(float ampliExp) {
        this.ampliExp = ampliExp;
    }

    @Override
    public @NotNull String toString() {
        return super.toString() + ", " +
                "TorsionLab{" +
                "weights='" + weights + '\'' +
                ", errWeights=" + errWeights +
                ", ampliTeo=" + ampliTeo +
                ", ampliExp=" + ampliExp +
                '}';
    }
}

File: app/src/main/java/uca/esi/manual/models/labs/TorsionLab.kt
package uca.esi.manual.models.labs

class TorsionLab(pId: String?) : BaseLab(pId, LabType.TORSION) {
    private val CARGAS_TORSION = intArrayOf(5, 10, 15, 20)
    private val MOMENTOS_TEO = floatArrayOf(.5f, 1.0f, 1.5f, 2.0f)
    private val AMPLI_TEO_TORSION = floatArrayOf(-.035f, -.073f, -.112f, -.15f)


    private var weights: String? = null
    private var errWeights = 0
    private var ampliTeo = 0f
    private var ampliExp = 0f

    init {
        setData()
    }

    override fun setData() {
        val index = (Math.random() * CARGAS_TORSION.size).toInt()
        weights = CARGAS_TORSION[index].toString()
        setValTeo(MOMENTOS_TEO[index])
        ampliTeo = AMPLI_TEO_TORSION[index]
    }

    override fun getData(): String? {
        return getWeights()
    }

    private fun getWeights(): String? {
        return weights
    }

    fun setWeights(weights: String?) {
        this.weights = weights
    }

    fun getErrWeights(): Int {
        return errWeights
    }

    fun setErrWeights(errWeights: Int) {
        this.errWeights = errWeights
    }

    fun getAmpliTeo(): Float {
        return ampliTeo
    }

    fun getAmpliExp(): Float {
        return ampliExp
    }

    fun setAmpliExp(ampliExp: Float) {
        this.ampliExp = ampliExp
    }
}